### PR TITLE
feat(backend): save `pk` JWT scope in our database

### PR DESCRIPTION
**Pull request by roschaefer**
Motivation
----------
This is a requirement for #2258.

We cannot control what Authentik sends us as a payload when we delete a user over there. What's included always is the `pk` attribute of a model. Therefore we must save the ID from Authentik in our database, too.

In general it is wise to encode a **permanent** and **immutable** ID in the JWT. We currently use `username`. But the `username` *is mutable* which will lead to problems in the future.

How to test
-----------
1. `cd backend`
2. `npm run test:unit`
3. Tests pass

! Important !
---------
~This PR **will crash our vServer** after deployment.~

~We must update the Authentik instance manually that is not part of our deployment cycle.~

~What needs to be done is the change in this [file](https://github.com/dreammall-earth/dreammall.earth/pull/2525/files#diff-ab0ffeecd111988f9570b4645447f2cf50e1aecc99eef4cc4d45fbfbe3314dcbR59) on line 59.~

@ulfgebhardt and I did this in a pair-programming.

Demo
---------
Here's how the database behaves:

1. On `origin/master` (8935c11627c7027e8c1a30e0519d519a96452ba1)
```
MariaDB [dreammall.earth]> Select * from User;
+-----+-------------------------+------------------------------+-------------------------+-----------+--------------+--------------+-------------+
| id  | username                | name                         | createdAt               | meetingId | availability | introduction | referenceId |
+-----+-------------------------+------------------------------+-------------------------+-----------+--------------+--------------+-------------+
|   1 | bibi                    | Bibi Bloxberg                | 2024-10-01 15:36:34.502 |      NULL | NULL         | NULL         | 95S7D245    |
|   2 | peter                   | Peter Lustig                 | 2024-10-01 15:36:34.502 |      NULL | NULL         | NULL         | VBPC5YN9    |

............................................................OMITTED...............................................................................

| 104 | Connor_Ziemann          | Frankie Mueller-Batz         | 2024-10-01 15:36:34.502 |      NULL | NULL         | NULL         | GAUSC24D    |
| 105 | akadmin                 | authentik Default Admin      | 2024-10-01 15:38:11.011 |      NULL | NULL         | NULL         | CKM9D1G8    |
+-----+-------------------------+------------------------------+-------------------------+-----------+--------------+--------------+-------------+
```

2. After running the database migration
```
MariaDB [dreammall.earth]> Select * from User;
+-----+-------------------------+------------------------------+-------------------------+-----------+--------------+--------------+-------------+------+
| id  | username                | name                         | createdAt               | meetingId | availability | introduction | referenceId | pk   |
+-----+-------------------------+------------------------------+-------------------------+-----------+--------------+--------------+-------------+------+
|   1 | bibi                    | Bibi Bloxberg                | 2024-10-01 15:36:34.502 |      NULL | NULL         | NULL         | 95S7D245    | NULL |
|   2 | peter                   | Peter Lustig                 | 2024-10-01 15:36:34.502 |      NULL | NULL         | NULL         | VBPC5YN9    | NULL |

............................................................OMITTED...............................................................................

| 104 | Connor_Ziemann          | Frankie Mueller-Batz         | 2024-10-01 15:36:34.502 |      NULL | NULL         | NULL         | GAUSC24D    | NULL |
| 105 | akadmin                 | authentik Default Admin      | 2024-10-01 15:38:11.011 |      NULL | NULL         | NULL         | CKM9D1G8    | NULL |
+-----+-------------------------+------------------------------+-------------------------+-----------+--------------+--------------+-------------+------+
```

3. After logging in for the first time:
```
MariaDB [dreammall.earth]> Select * from User;
+-----+-------------------------+------------------------------+-------------------------+-----------+--------------+--------------+-------------+------+
| id  | username                | name                         | createdAt               | meetingId | availability | introduction | referenceId | pk   |
+-----+-------------------------+------------------------------+-------------------------+-----------+--------------+--------------+-------------+------+
|   1 | bibi                    | Bibi Bloxberg                | 2024-10-01 15:36:34.502 |      NULL | NULL         | NULL         | 95S7D245    | NULL |
|   2 | peter                   | Peter Lustig                 | 2024-10-01 15:36:34.502 |      NULL | NULL         | NULL         | VBPC5YN9    | NULL |

............................................................OMITTED...............................................................................

| 104 | Connor_Ziemann          | Frankie Mueller-Batz         | 2024-10-01 15:36:34.502 |      NULL | NULL         | NULL         | GAUSC24D    | NULL |
| 105 | akadmin                 | authentik Default Admin      | 2024-10-01 15:38:11.011 |      NULL | NULL         | NULL         | CKM9D1G8    |    4 |
+-----+-------------------------+------------------------------+-------------------------+-----------+--------------+--------------+-------------+------+
```

image

### CODE REVIEW

File: backend/prisma/migrations/20241001140736_add_pk_to_users/migration.sql [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/*
+  Warnings:
+
+  - A unique constraint covering the columns `[pk]` on the table `User` will be added. If there are existing duplicate values, this will fail.
+  - Added the required column `pk` to the `User` table without a default value. This is not possible if the table is not empty.
+
+*/
+-- AlterTable
+ALTER TABLE `User` ADD COLUMN `pk` INTEGER NOT NULL;
```
Comment (ulfgebhardt) [inline backend/prisma/migrations/20241001140736_add_pk_to_users/migration.sql:9]: I think we might consider the id to be a string not an integer, considering that other pk-ids behave like uuids.

image
https://docs.goauthentik.io/docs/events/

Furthermore I believe that a not null constraint will result in migration problems.

File: backend/src/context/context.spec.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+      beforeEach(() => {
+        const brokenPayload = {
+          key: { type: 'whatever' },
+          payload: { nickname: 'nickname', name: 'name' },
+          protectedHeader: { alg: 'RS256' },
+        } satisfies Awaited<ReturnType<typeof jwtVerify<{ nickname: string; name: string }>>>
+        jest.mocked(jwtVerify<{ nickname: string; name: string }>).mockResolvedValue(brokenPayload)
+      })
+
+      it('crashes quite dramatically', async () => {
```
Comment (roschaefer) [inline backend/src/context/context.spec.ts:100]: This is what happens if we forget our Authentik at `auth.dreammall.earth`.

File: backend/prisma/migrations/20241001140736_add_pk_to_users/migration.sql (hunk 1 above)
Comment (ulfgebhardt) [reply backend/prisma/migrations/20241001140736_add_pk_to_users/migration.sql]: Its fine with me to leave it like it is

**Review by ulfgebhardt — CHANGES_REQUESTED** [review]
Just doing `docker compose up database authentik backend frontend`

results in 
image

This results in a broken frotnend 
image


`docker compose up database authentik backend frontend ---build` does not solve the problem
image

I'd like to have a solution for the developers before merging this

**Review by ulfgebhardt — APPROVED** [review]

**Pull request merged**